src/OrleansServiceBus/Providers/Streams/EventHub/EventHubDataAdapter.cs

```diff
@@ -16,13 +17,49 @@ public struct CachedEventHubMessage
     {
         public Guid StreamGuid;
         public long SequenceNumber;
+        public DateTime EnqueueTimeUtc;
+        public DateTime DequeueTimeUtc;
         public ArraySegment<byte> Segment;
     }
 
     /// <summary>
+    /// Replication of EventHub EventData class, reconstructed from cached data CachedEventHubMessage
+    /// </summary>
+    [Serializable]
+    public class EventHubMessage
+    {
+        public EventHubMessage(ref CachedEventHubMessage cachedMessage)
+        {
+            int readOffset = 0;
+            StreamIdentity = new StreamIdentity(cachedMessage.StreamGuid, SegmentBuilder.ReadNextString(cachedMessage.Segment, ref readOffset));
+            SequenceNumber = cachedMessage.SequenceNumber;
+            EnqueueTimeUtc = cachedMessage.EnqueueTimeUtc;
+            DequeueTimeUtc = cachedMessage.DequeueTimeUtc;
+            Properties = SegmentBuilder.ReadNextBytes(cachedMessage.Segment, ref readOffset).DeserializeProperties();
+            object offsetObj;
+            Offset = Properties.TryGetValue("Offset", out offsetObj)
```

Comment:
In the EventData code, all of its properties (PartitionKey, SequenceNumber, Offset, ...) all come from the property bag, and when modified are set in the property bag.  That is, the property bag is the authoritative source of these values.
I considered following that pattern here, but since these properties are read-only, I decided to set the common properties for performance, rather than get them from the bag each time.